docs/rfcs/functional-composition-pattern.md

```diff
@@ -0,0 +1,311 @@
+New<Type>(<Method1>Func, <Method2>Func, ...) Type` for all methods,
+using the concrete implementation struct. For example:
+
+```go
+func NewRateReservation(wf WaitTimeFunc, cf CancelFunc) RateReservation {
+    return rateReservationImpl{
+        WaitTimeFunc: wf,
+        CancelFunc:   cf,
+    }
+}
+
+func NewRateLimiter(f ReserveRateFunc) RateLimiter {
+    return rateLimiterImpl{ReserveRateFunc: f}
+}
+```
+
+[The Go language automatically converts function literal
+expressions](https://go.dev/doc/effective_go#conversions) into the
+correct named type (i.e., `<Method>Func`), so we can pass function
+literals to these constructors without an explicit conversion:
+
+```go
+    return NewRateReservation(
+        // Wait time 1 second
+        func() time.Duration { return time.Second },
+	// Cancel is a no-op.
+	nil,
+    )
+```
+
+For more complicated interfaces, this pattern can be combined with the
+[Functional Option
+pattern](https://commandcenter.blogspot.com/2014/01/self-referential-functions-and-design.html)
+in Golang, shown in the next example.
+
+Taken from `receiver/receiver.go`, here we setup signal-specific
+functions using a functional-option argument passed to
+`receiver.NewFactory`:
+
+```go
+// Setup optional signal-specific functions (e.g., logs)
+func WithLogs(createLogs CreateLogsFunc, sl component.StabilityLevel) FactoryOption {
+	return factoryOptionFunc(func(o *factoryImpl, cfgType component.Type) {
+		o.CreateLogsFunc = createLogs
+		o.LogsStabilityFunc = sl.Self // See (5) below
+	})
+}
+
+// Accept options to configure various aspects of the interface
+func NewFactory(cfgType component.Type, createDefaultConfig component.CreateDefaultConfigFunc, options ...FactoryOption) Factory {
+	f := factoryImpl{
+		Factory: component.NewFactory(cfgType.Self, createDefaultConfig),
+	}
+	for _, opt := range options {
+		opt.applyOption(&f, cfgType)
+	}
+	return f
+}
+```
+
+### 4. Seal Public Interface Types
+
+Using an unexported method "seals" the interface type so external
+packages can only use, not implement the interface. This allows
+interfaces to evolve safely because users are forced to use
+constructor functions.
+
+```go
+// Public interfaces must include at least one private method
+type RateLimiter interface {
+    ReserveRate(context.Context, int) (RateReservation, error)
+
+    // Prevents external implementations
+    private()
+}
+
+// Concrete implementations are sealed with this method.
+type rateLimiterImpl struct {
+    ReserveRateFunc
+}
+
+func (rateLimiterImpl) private() {}
+```
+
+This practice enables safely evolving interfaces. A new method can be
+added to a public interface type because public constructor functions
+force the user to obtain the new type and the new type is guaranteed
+to implement the old interface. If the functional option pattern is
+already being used, then new interface methods will not require new
+constructors, only new options. If the functional option pattern is
```

Comment:
I think as mentioned in previous comments, we should enforce the Optional pattern even if we don't have any option at the beginning.

Reply:
I see. I read this like "new public interfaces SHOULD use functional options in their constructors" for maximum forward compatibility, b/c it means we can always add new options w/o a breaking change".